In [104]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report
import matplotlib.pyplot as plt

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [138]:
# Read the CSV file from the Resources folder into a Pandas DataFrame

# Read the CSV file
lending_data = pd.read_csv('Resources/lending_data.csv')

# Display the first few rows of the dataframe
lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [120]:
# Remove rows with missing values
lending_data_cleaned = lending_data.dropna()

In [121]:
# Review the DataFrame
lending_data_cleaned.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%           3.000000          0.000000  14800.000000      0.000000  
50%           4.000000          0.000000  18100.000000      0.000000  
75%           4.000000          1.000000  21400.000000      0.000000  
max          16.000000          3.000000  75200.000000      1.000000

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [122]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_cleaned["loan_status"]

# Separate the X variable, the features
X = lending_data_cleaned.drop(columns=["loan_status"])

In [123]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [124]:
# Review the X variable DataFrame
X.head() 

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [125]:
# Review the X variable DataFrame
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
dtypes: float64(3), int64(4)
memory usage: 4.1 MB


In [126]:
# Review the X variable DataFrame
X.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  
count     77536.000000      77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804  
std           1.904426          0.582086   8371.635077  
min           0.000000          0.000000      0.000000  
25%           3.000000          0.000000  14800.000000  
50%           4.000000          0.000000  18100.000000  
75%           4.000000          1.000000  21400.000000  
max          16.000000          3.000000  75200.000000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [127]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split


# Split the data into training and testing sets with random_state=1
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y, 
                                                    test_size=0.2, 
                                                    random_state=1)


In [129]:
# Display the shapes of the resulting sets
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (62028, 7)
X_test shape: (15508, 7)
y_train shape: (62028,)
y_test shape: (15508,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [130]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the logistic regression model with random_state=1
model = LogisticRegression(solver='lbfgs', random_state=1)

# Train / Fit the model using the training data
model.fit(X_train, y_train)

LogisticRegression(random_state=1)

# Validate the model using the test data

In [131]:
# Score the model
print(f"Training Data Score: {model.score(X_train, y_train)}")
print(f"Testing Data Score: {model.score(X_test, y_test)}")

Training Data Score: 0.9925356290707422
Testing Data Score: 0.9930358524632448


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [132]:
# Make a prediction using the testing data
y_pred = model.predict(X_test)
pd.DataFrame({"Prediction": y_pred, "Actual": y_test})

Prediction  Actual
60914           0       0
36843           0       0
1966            0       0
70137           0       0
27237           0       0
...           ...     ...
30965           1       0
30996           0       0
69036           0       0
23886           0       0
59106           0       0

[15508 rows x 2 columns]

In [133]:
# Display the first few predictions
print(y_pred[:10])

[0 0 0 0 0 0 0 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [134]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, y_pred)

In [135]:
# Print the classification report for the model
# Generate the classification report
report = classification_report(y_test, y_pred)

# Print the classification report
print(report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** WRITE YOUR ANSWER HERE!
The classification report provides metrics for performance evaluation of the logistic regression model. 

A.  Class Specific Performance 
    
    Class 0 (Healthy loan)
    * Precision 1.00 which is 100% - All predicted positives for class 0, 100% were correct
    * Recall comes as 99% - out of all actual positives for class 0, 99% were correctly identified by the model
    * F1-Score - 100% showing very high balance between precision and recall. 


    Class 1 (High-risk loan)
    * Precision 86%, recall 94% and F1 Score 90%  were correctly identified by the model. 

B.  Average Metrics
    
    Macro Average 
    * Precision 93%, reall 97% and F1-score 95% - noted as average over both classes, treating each class equally. 

    Weightage Agerage 
    *  Precision 99%, recall 99% and F1 score 99% - noted as average over the both classes, weighted by the number of instances in both classes.

In summary, the logistic regression model shows excellent overall performance with an accuracy of 99%.  The model performs exceptionally well in predicting class 0 (Healthy loan) with perfect precision and very high recall. For class 1 (High-risk loan), the model also performs well however there are some rooms for improvement in precision. Finally the high weighted average metrics indicate that the model is reliable across the dataset though slight drop in precision for class 1 suggesting model occasionally misclassifies some loan approvals. 





---